### PR TITLE
BUG: Fix pinv for stacked matrices

### LINKED ISSUE #8826: BUG: Linalg.pinv fails on stacks of matrices
```python
>>> a = np.stack((np.eye(3),)*4, axis=0)
>>> ai = np.linalg.inv(a)
>>> assert (a == ai).all()
>>> api = np.linalg.pinv(a)
Traceback (most recent call last):
  File "<pyshell#9>", line 1, in <module>
    np.linalg.pinv(a)
  File "C:\Program Files\Python 3.5\lib\site-packages\numpy\linalg\linalg.py", line 1668, in pinv
    if s[i] > cutoff:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes #8826

Duplicate of #4678

### CODE REVIEW

File: numpy/linalg/tests/test_linalg.py [hunk 1]
```diff
@@ -423,25 +423,6 @@ def test_generalized_empty_herm_cases(self):
             exclude={'none'})
 
 
-def dot_generalized(a, b):
```
Comment (eric-wieser) [inline numpy/linalg/tests/test_linalg.py]: Ahead of its time...

File: numpy/linalg/linalg.py [hunk 2]
```diff
@@ -226,6 +227,22 @@ def _assertNoEmpty2d(*arrays):
         if _isEmpty2d(a):
             raise LinAlgError("Arrays cannot be empty")
 
+def transpose(a):
+    """
+    Transpose each matrix in a stack of matrices.
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: TODO: expose this publicly?

Comment (mhvk) [inline numpy/linalg/linalg.py]: Well, there is at least one implementation that could (eventually) be abandoned if this is exposed: https://github.com/astropy/astropy/blob/master/astropy/coordinates/matrix_utilities.py#L32

Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Any reason for calling the method and not the free function?

File: numpy/linalg/tests/test_linalg.py (hunk 1 above)
Comment (eric-wieser) [inline numpy/linalg/tests/test_linalg.py]: Except not, as it turns out. This doesn't broadcast, but instead tries to work out which are vectors and which are matrices

File: numpy/linalg/linalg.py [hunk 3]
```diff
@@ -2008,13 +2020,13 @@ def lstsq(a, b, rcond=-1):
                 resids = array([sum((ravel(bstar)[n:])**2)],
                                dtype=result_real_t)
     else:
-        x = array(transpose(bstar)[:n,:], dtype=result_t, copy=True)
+        x = array(nd_transpose(bstar)[:n,:], dtype=result_t, copy=True)
```
Comment (mhvk) [inline numpy/linalg/linalg.py]: If you're changing it anyway, maybe just do `bstar.T`? So much clearer...

Comment (mhvk) [inline numpy/linalg/linalg.py]: Of course, since `bstar` is 2-D, you can also just leave it as it was...

Comment (eric-wieser) [inline numpy/linalg/linalg.py]: I mean, I plan to throw this all out in #8720 anyway...

File: numpy/linalg/tests/test_linalg.py (hunk 1 above)
Comment (mhvk) [inline numpy/linalg/tests/test_linalg.py]: Funny, I had exactly the same thought -- couldn't you use `matmul` instead -- but then realised it was actually different (in the actual expression, I might have used `einsum` instead, but it is not really important at all).

File: numpy/linalg/linalg.py (hunk 2 above)
Comment (mhvk) [inline numpy/linalg/linalg.py]: The method is the real thing; the function just turns the input into an array if needed and then calls the method. https://github.com/numpy/numpy/blob/master/numpy/core/fromnumeric.py#L458

Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Yep, I know :). Makes it break on duct-types though.

Comment (mhvk) [inline numpy/linalg/linalg.py]: p.s. In the astropy case this is only used internally in cases where I *know* the inputs are arrays; if we exposed this publically in numpy, I think using the function makes more sense.

File: numpy/linalg/linalg.py (hunk 3 above)
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Done

File: doc/release/1.13.0-notes.rst [hunk 4]
```diff
@@ -159,6 +159,10 @@ being iterated over.
 For consistency with ``ndarray`` and ``broadcast``, ``d.ndim`` is a shorthand
 for ``len(d.shape)``.
 
+``matrix_rank`` and ``pinv`` in ``np.linalg`` now works for stacked arrays
+--------------------------------------------------------------------------
```
Comment (charris) [inline doc/release/1.13.0-notes.rst]: works -> work

File: numpy/linalg/linalg.py [hunk 5]
```diff
@@ -1615,24 +1632,25 @@ def pinv(a, rcond=1e-15 ):
...
-      Cutoff for small singular values.
-      Singular values smaller (in modulus) than
-      `rcond` * largest_singular_value (again, in modulus)
-      are set to zero.
+    a : (..., M, N) array_like
+        Matrix to be pseudo-inverted. The ability to handle stacked
+        matrices was added in Numpy 1.13.0.
+    rcond : (...) arr_like of float
+        Cutoff for small singular values.
+        Singular values smaller (in modulus) than
```
Comment (charris) [inline numpy/linalg/linalg.py]: How about "less than" rather than smaller. The singular values are always non-negative reals.

Comment (charris) [inline numpy/linalg/linalg.py]: Even for complex matrices.

Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Note that I haven't actually changed these comments, just adjusted their indentation.

> The singular values are always non-negative reals.

Is this the case even under numerical error?

Either way, the description right now matches the implementation. If taking the modulus is not required, then we should remove it from the doc and implementation at the same time (and in another PR).

Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Note that "modulus" was added here in d1572e28dd3ebaee82d747cc20eca7e86ade52d3 by @pv, so looks like an intentional clarification

Comment (mhvk) [inline numpy/linalg/linalg.py]: Puzzling -- as @charris notes, singular values are by definition positive real, so reading something about a modulus would seem only confusing.

**Comment by eric-wieser** [comment]
Updated to refer to 1.14 not 1.13 in release notes

**Comment by charris** [comment]
@eric-wieser Want to rebase this? I'm trying to get all the linalg fixes/enhancements in.

**Comment by eric-wieser** [comment]
Yes, but I can't until I regain access to a git installation in two weeks - so maybe someone else should.

**Comment by eric-wieser** [comment]
Conflict is just in the imports

**Comment by eric-wieser** [comment]
@charris: Ok, rebased. Ready to go in?

File: numpy/linalg/linalg.py (hunk 2 above)
Comment (charris) [reply numpy/linalg/linalg.py]: We need something like this, but I think it should be an array method as well. Maybe something like `a.M(atrix)T(ranspose`. When the `*.T` method was proposed there was discussion of what it should do, transposing matrices in a stack was one of ideas. I could also see a vector transpose, `a.VT` which just appends 1 to the dimensions. These ideas need discussion on the list.

Comment (eric-wieser) [reply numpy/linalg/linalg.py]: Should I add a leading underscore to the name for now?

Comment (charris) [reply numpy/linalg/linalg.py]: @eric-wieser That would help distinguish it from `np.transpose` and avoid some confusion. Long term, it probably needs a different name.

**Comment by charris** [comment]
I'm putting this in as is. Thanks Eric. If you want to change the name of `transpose` go ahead and make another PR.

File: numpy/linalg/linalg.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/linalg/linalg.py]: At any rate, this isn't going to appear in the docs at the moment, so isn't really public API yet.

I'll send something around the mailing list requesting a name at some point. I think we can start off with the conservative `assert ndim >= 2` semantics initially, and put off the decision in #7495 and #9530

**Pull request merged**